src/execution/operator/aggregate/physical_window.cpp

```diff
@@ -211,253 +162,248 @@ class WindowGlobalSinkState : public GlobalSinkState {
-	idx_t count;
-};
+	// If the local partition is now too small, flush it and reallocate
+	auto new_partition = grouping_data->CreateShared();
+	auto new_append = make_unique<PartitionedColumnDataAppendState>();
+	new_partition->InitializeAppendState(*new_append);
+
+	local_partition->FlushAppendState(*local_append);
+	auto &local_groups = local_partition->GetPartitions();
+	for (auto &local_group : local_groups) {
+		ColumnDataScanState scanner;
+		local_group->InitializeScan(scanner);
 
-bool WindowLocalHashGroup::SinkChunk(DataChunk &sort_buffer, DataChunk &input_chunk) {
-	D_ASSERT(sort_buffer.size() == input_chunk.size());
-	count += input_chunk.size();
-	auto &global_sort = *global_group.global_sort;
-	if (!local_sort) {
-		local_sort = make_unique<LocalSortState>();
-		local_sort->Initialize(global_sort, global_sort.buffer_manager);
+		DataChunk scan_chunk;
+		local_group->InitializeScanChunk(scan_chunk);
+		for (scan_chunk.Reset(); local_group->Scan(scanner, scan_chunk); scan_chunk.Reset()) {
+			new_partition->Append(*new_append, scan_chunk);
+		}
 	}
 
-	local_sort->SinkChunk(sort_buffer, input_chunk);
+	// The append state has stale pointers to the old local partition, so nuke it from orbit.
+	new_partition->FlushAppendState(*new_append);
 
-	if (local_sort->SizeInBytes() >= global_group.memory_per_thread) {
-		local_sort->Sort(global_sort, true);
+	local_partition = std::move(new_partition);
+	local_append = make_unique<PartitionedColumnDataAppendState>();
+	local_partition->InitializeAppendState(*local_append);
+}
+
+void WindowGlobalSinkState::UpdateLocalPartition(GroupingPartition &local_partition, GroupingAppend &local_append) {
+	// Make sure grouping_data doesn't change under us.
+	lock_guard<mutex> guard(lock);
+
+	if (!local_partition) {
+		local_partition = grouping_data->CreateShared();
+		local_append = make_unique<PartitionedColumnDataAppendState>();
+		local_partition->InitializeAppendState(*local_append);
+		return;
 	}
 
-	return (local_sort->SizeInBytes() >= global_group.memory_per_thread);
+	// 	Grow the groups if they are too big
+	ResizeGroupingData(count);
+
+	//	Sync local partition to have the same bit count
+	SyncLocalPartition(local_partition, local_append);
 }
 
-void WindowLocalHashGroup::Combine() {
-	if (local_sort) {
-		global_group.Combine(*local_sort);
-		global_group.count += count;
-		local_sort.reset();
+void WindowGlobalSinkState::CombineLocalPartition(GroupingPartition &local_partition, GroupingAppend &local_append) {
+	if (!local_partition) {
+		return;
 	}
+	local_partition->FlushAppendState(*local_append);
+
+	// Make sure grouping_data doesn't change under us.
+	// Combine has an internal mutex, so this is single-threaded anyway.
+	lock_guard<mutex> guard(lock);
+	SyncLocalPartition(local_partition, local_append);
+	grouping_data->Combine(*local_partition);
+}
+
+void WindowGlobalSinkState::BuildSortState(ColumnDataCollection &group_data, WindowGlobalHashGroup &hash_group) {
```

Comment:
Do the other threads wait while one thread executes this function? I am trying to understand the flow of things here. Lets discuss this later

Reply:
It just builds the sort for a single group, so different threads can build the sort for other partitions. But it would be nice if the sort construction was multithreaded too.

(The code is in `WindowGlobalSinkState` because it uses some other global state but maybe it could be moved to be less confusing?)